### PR TITLE
PR on all of HW2 code

### PR DESCRIPTION
**Pull request by keviphan**
### Description of changes made
I added my implementation for each of the homework 2's given JavaScript file, 01, 02, 03, and 04.

### Reason for changes
I made these changes to complete hw2 requirements

### Feedback Request (by Wednesday 2/5/25)
Check my work on logic implementation for 01
Check if I missed any readability points and any airbnb style guides
*note, for 03 - for the welcome route, I get a 304 status code, not sure why, instructions say it needs to be 200
*note, I used plain text for the cache route for both 03/04, I followed the pdf version of the instructions and it said to use plain text which differs from the instructions within the files itself.
Feel free to suggest anything else, thank you.

### CODE REVIEW

**Comment by eyefremov** [comment]
# 01-coins.js:
- You should add a check for negative values, testing a negative input of -0.01 yielded an incorrect output of:
```
3 quarters, 2 dimes, 4 pennies
```
- Add a check to see if it even is a number; numbers passed in even as a string are still valid, but an input of "Hello" yields nothing as an output. If that's what you desire, then the advice is unnecessary. You can test this with the code below.
```js
console.log(calculateChange('Hello'));
```
- You can potentially add more test cases to see how the code behaves. Naturally, the code passes the two tests that are present already.

# 02-node-routing.js:
- [Line 47](https://github.com/keviphan/kevin-phan-fullstack/pull/2/files#diff-5655bdb7a2b27c8c74782f68f31077d6a627bd998f61c68542121c60cc0ad227R47): slight spelling error, "redrected" -> "redirected"
```js
res.write('<h1>You have been redrected to the /redirected</h1>');
```
- [Line 57](https://github.com/keviphan/kevin-phan-fullstack/pull/2/files#diff-5655bdb7a2b27c8c74782f68f31077d6a627bd998f61c68542121c60cc0ad227R57): The instructions ask for returning in HTML format, but you haven't done so.
#### Before:
```js
res.write('This resource was cached');
```
#### After:
```js
res.write('<p>This resource was cached</p>'); // Or use <h1> if that's what you prefer.
```
- [Lines 15-22](https://github.com/keviphan/kevin-phan-fullstack/pull/2/files#diff-5655bdb7a2b27c8c74782f68f31077d6a627bd998f61c68542121c60cc0ad227R15-R22): the routes variable is never used. I don't know about removing it since I'm unsure if using the variable is required.
```js
  const routes = [
    'welcome',
    'redirect',
    'redirected',
    'cache',
    'cookie',
    'other',
  ];
```

# 03-express-routing.js:
- For your welcome route returning a status code of 302, this could be standard express.js behavior that optimizes caching after you visit the route more than once. In other words, your first visit will result in a status code of 200, and subsequent visits will result in a code of 302. You can try to disable this behavior with:
```
app.disable('etag');
```
-  [Line 50:](https://github.com/keviphan/kevin-phan-fullstack/pull/2/files#diff-a63427156b15ba2691cd08f35451312d6ba4e0360355534e5f1019b00a5198ecR50) Same issue as 02-node-routing where you misspell redirecting.
#### Before:
```js
res.send('<h1>You have been redrected to the /redirected</h1>');
```
#### After:
```js
res.send('<h1>You have been redirected to the /redirected</h1>');
```
# 04-form.js
- [Lines 18-21](https://github.com/keviphan/kevin-phan-fullstack/pull/2/files#diff-4a00c468c54f1e222cc5463a1db62b60cdabd03577617ee2f10604e1c6d0e0bdR18-R21): You can make your code a bit more concise:
#### Before:
```js
app.get('/', (req, res) => {
  res.status(200);
  res.redirect('/form');
});
```
#### After:
```js
app.get('/', (req, res) => {
  res.redirect(302, '/form'); // Although it will default to 302 anyways, so you could remove the status code as well.
});
```
- [Lines 63-67](https://github.com/keviphan/kevin-phan-fullstack/pull/2/files#diff-4a00c468c54f1e222cc5463a1db62b60cdabd03577617ee2f10604e1c6d0e0bdR63-R67): Not an error, but you could apply chaining here:
#### Before:
```js
app.use((req, res) => {
  res.status(404);
  res.set({ 'Content-Type': 'text/html' });
  res.send('<h1>404: Page not found</h1>');
});
```
#### After:
```js
app.use((req, res) => {
  res
    .status(404)
    .set('Content-Type', 'text/html')
    .send('<h1>404: Page not found</h1>');
});
```

File: hw2/01-coins.js [hunk 1]
```diff
@@ -1,6 +1,33 @@
 /** Exercise 01 - Coins **/
 
-// Add your function here 
+// Add your function here
+const calculateChange = (n) => {
+  if (n > 100) return 'the number is too large';
+
```
Comment (Ghaadah) [inline hw2/01-coins.js:17]: adding more checking for the input cases will improve the code.

File: hw2/04-form.js [hunk 2]
```diff
@@ -1,5 +1,71 @@
...
+
+// main route --> /
+// (redirect to form)
+app.get('/', (req, res) => {
+  res.status(200);
+  res.redirect('/form');
+});
+
+// form route --> /form
+app.get('/form', (req, res) => {
```
Comment (Ghaadah) [inline hw2/04-form.js:23]: we may use multiple routes in the same request. Instead of redirecting the main route / to /form (lines 18 -21), you may update line 24 to include array of routes.  
app.get(['/','/form],(req, res)=>{

I think this will make the code cleaner.

**Review by Ghaadah — COMMENTED** [review]
Hi Kevin,
Great Code, Here are some suggestion for your review:

01- I suggest to add more input validation cases .(after line 5).  The code is clean, readable and functional. Great Code!

02  and 03 -  I tested the code and all routes were working perfectly. During the ESLint test, I encountered only one error, an unused const routes. I know it was part of the original homework starter and I didn't remove it either, but it looks like we should delete it since it's not being used. 

04- Looks great! I left a comment in the file change for your review. After going through your code, I realized I might have missed some parts for this section. I followed the starter file and only added the name and email, but it looks like we also need the comment and checkbox fields.
